feature(remap source): Emitting multiple events

jszwedko opened the pull request:
Few remaining TODOs, but this is almost there. The approach I took was to introduce an "adapter" for `vector::Event` rather than having `event::Metric` and `event::LogEvent` implement `vrl::Target` directly. This allowed me to decouple the implementation for `LogEvent` from the underlying data structure; instead taking the shortcut of just representing it as a `vrl::Value`, which already implements `vrl::Target`, and then marshaling back into a `Vec<Event>` to map arrays.

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

<!--
**Your PR title must conform to the conventional commit spec!**

  <type>!?(<scope>): <description>

  * `type` = chore, enhancement, feat, fix
  * `!` = signals a breaking change
  * `scope` = https://github.com/timberio/vector/blob/master/.github/semantic.yml#L4
  * `description` = short description of the change

Examples:

  * enhancement(file source): Added `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fixed a bug discovering new files
  * chore(external docs): Clarified `batch_size` option
-->

jszwedko commented on src/event/log_event.rs:
```diff
@@ -473,259 +408,6 @@ mod test {
         assert_eq!(json.pointer("/map/map/non-existing"), Some(&fallback));
     }
 
-    #[test]
```
These tests are no longer needed as we just proxy to a `vrl::Value` for `LogEvent` instead.

jszwedko replied:
Update: re-added these tests in `vrl_target.rs`.

jszwedko commented on src/event/metric.rs:
```diff
@@ -1174,137 +1013,4 @@ mod test {
         );
     }
-
-    #[test]
```
These tests I would like to move to `vrl_target.rs`.

pablosichert commented on src/event/vrl_target.rs:
```diff
@@ -0,0 +1,398 @@
+impl VrlTarget {
+    pub fn new(event: Event) -> Self {
+        match event {
+            // TODO optimize
```
Do you have anything specific in mind here?

jszwedko replied:
I was able to take advantage of the existing `From` implementations in 3cf40a339 but I have a feeling this is doing more copying than we might like. I plan to run a benchmark or two to see if the overhead is noticeable. I'm curious to hear any suggestion though.

pablosichert replied:
Ah, I see. So the concern is having three steps of `.into` conversions instead of (if possible) one complete conversion?

jszwedko replied:
Well, it's more that the last `into` (`VrlTarget::LogEvent(value.into())`) ends up doing a `BTreeMap` clone to map the values. This appears to be where Vector ends up spending a bunch of time in my benches.

jszwedko replied:
I was able to resolve this overhead by, instead of storing a `vrl::Value` in my new `vrl::Target::Log` implementation, storing a `event::Value`. This avoids needing to do the btreemap copy to `vrl::Value` on instantiation and back to `event::Value` when converting back into a `LogEvent`.

pablosichert commented on src/event/vrl_target.rs:
```diff
@@ -0,0 +1,398 @@
+    pub fn into_events(self) -> impl Iterator<Item = Event> {
+        match self {
+            VrlTarget::LogEvent(value) => {
+                Box::new(value_into_events(value)) as Box<dyn Iterator<Item = Event>>
```
Is the explicit `Box<dyn Iterator<Item = Event>>` needed because the compiler chooses a too specific type for each arm?

jszwedko replied:
Yeah, that's exactly it :)

jszwedko replied:
`as Box<dyn _>` returns a compiler error:

```
error: `<` is interpreted as a start of generic arguments for `Box`, not a comparison
  --> src/event/vrl_target.rs:48:72
   |
48 |                 Box::new(std::iter::once(Event::Metric(metric))) as Box<dyn _>
   |                 -------------------------------------------------------^----- interpreted as generic arguments
   |                 |                                                      |
   |                 |                                                      not interpreted as comparison
   |                 help: try comparing the cast value: `(Box::new(std::iter::once(Event::Metric(metric))) as Box)`

error: expected expression, found keyword `dyn`
  --> src/event/vrl_target.rs:48:73
   |
48 |                 Box::new(std::iter::once(Event::Metric(metric))) as Box<dyn _>
   |                                                                         ^^^ expected expression
```

pablosichert commented on src/event/vrl_target.rs:
```diff
@@ -0,0 +1,398 @@
+        vrl::Value::Array(values) => Box::new(values.into_iter().map(value_into_events).flatten())
+            as Box<dyn Iterator<Item = Event>>,
+        vrl::Value::Object(object) => {
+            // TODO optimize
```
Same question here.

pablosichert commented on docs/reference/components/transforms/remap.cue:
```diff
@@ -138,6 +138,18 @@ components: transforms: "remap": {
 		}
+		emitting_multiple_events: {
+			title: "Emitting multiple log events"
+			body: #"""
+				Multiple log events can be emitted from remap by setting the
+				root path, `.` to an array. One log event will be emitted for
+				each input array element.
+
+				If any of the array elements are not an object, they will be set
+				as the set as the `message` key for the corresponding emitted
+				log event.
+				"""#
```
Some slight rewording / corrections:

```suggestion
			body: #"""
				Multiple log events can be emitted from remap by assigning an
				array to the root path `.`. One log event will be emitted for
				each input element of the array.

				If any of the array elements is not an object, they will be set
				as the `message` key for the corresponding emitted log event.
				"""#
```

I didn't understand the second sentence without looking at the implementation.

pablosichert commented on src/transforms/remap.rs:
```diff
@@ -167,6 +175,37 @@ mod tests {
+        let metadata = event.metadata().clone();
+
+        let conf = RemapConfig {
+            source: r#"
+. = .events
+"#
```
Good place to use `indoc` here

pablosichert commented on src/event/vrl_target.rs:
```diff
@@ -0,0 +1,395 @@
+        }
+    }
+
+    /// Turn the target back into events
+    ///
+    /// This returns an iterator of events as one event can be turned into multiple by assign `.`
+    /// to an array in VRL
```
```suggestion
    /// Turn the target back into events.
    ///
    /// This returns an iterator of events as one event can be turned into multiple events by
    /// assigning an array to `.` in VRL.
```

jszwedko replied:
👍 we seem to have a mix of doc comments with and without `.`s. I wonder if there is a way we could enforce `.`s if we prefer that. I don't have a strong opinion.

pablosichert commented on src/event/vrl_target.rs:
```diff
@@ -0,0 +1,395 @@
+/// We can get the `type` of the metric in Remap, but can't set  it.
+const VALID_METRIC_PATHS_GET: &str = ".name, .namespace, .timestamp, .kind, .tags, .type";
+
+/// An adapter to turn `Event`s into `vr::Target`s
```
```suggestion
/// An adapter to turn `Event`s into `vrl::Target`s.
```

pablosichert commented on src/event/vrl_target.rs:
```diff
@@ -0,0 +1,395 @@
+
+const VALID_METRIC_PATHS_SET: &str = ".name, .namespace, .timestamp, .kind, .tags";
+
+/// We can get the `type` of the metric in Remap, but can't set  it.
```
```suggestion
/// We can get the `type` of the metric in Remap, but can't set it.
```

StephenWakely commented:
This is sweet! 

It will be very interesting to benchmark this as per https://github.com/timberio/vector/pull/5374#issuecomment-816314169. I suspect it will do wonders for a Vrl heavy config since from what I can gather we are no longer paying the cost of converting between a Vrl Value and a Vector Value with each call.

The multiple events is just an added bonus!

jszwedko commented:
I ran a few benchmarks and found some interesting results:

* It seems like inserting, via the `Target` implementation, is generally faster for `event::LogEvent` than it is `vrl::Value`. The more `insert`s I do, the bigger the gap between master and this branch.  I'm curious if you have thoughts there @StephenWakely . There could be something else I'm doing here that is causing that behavior though.

Benchmarks using a config that looks like:

```toml
[sources.in]
type = "stdin"

[transforms.remap]
type = "remap"
inputs = ["in"]
drop_on_error = true
source = """
.foo0 = "bar"
# adding additional inserts here
"""

[sinks.out]
type = "blackhole"
inputs = ["remap"]
```

Master, empty remap transform:

```
 Performance counter stats for 'bash -c yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ./target/release/vector-master-0240ae601 --config ~/tmp/remap.toml > /dev/null' (10 runs):

      25976.082806      task-clock:u (msec)       #    2.533 CPUs utilized            ( +-  0.21% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              5388      page-faults:u             #    0.207 K/sec                    ( +-  2.80% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      10.256256539 seconds time elapsed                                          ( +-  0.21% )
```

This branch, empty remap transform:

```
 Performance counter stats for 'bash -c yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ./target/release/vector-emit-multiple-3ca72b111 --config ~/tmp/remap.toml > /dev/null' (10 runs):

      30524.937509      task-clock:u (msec)       #    2.744 CPUs utilized            ( +-  0.57% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              4951      page-faults:u             #    0.162 K/sec                    ( +-  0.33% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      11.123577205 seconds time elapsed                                          ( +-  0.55% )
```

A roughly 8% penalty.

Master, one field assignment:

```
 Performance counter stats for 'bash -c yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ./target/release/vector-master-0240ae601 --config ~/tmp/remap.toml > /dev/null' (10 runs):

      27542.389545      task-clock:u (msec)       #    2.626 CPUs utilized            ( +-  0.55% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              5251      page-faults:u             #    0.191 K/sec                    ( +-  2.67% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      10.488638371 seconds time elapsed                                          ( +-  0.78% )
```

This branch, one field assignment:

```
 Performance counter stats for 'bash -c yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ./target/release/vector-emit-multiple-3ca72b111 --config ~/tmp/remap.toml > /dev/null' (10 runs):

      33143.892345      task-clock:u (msec)       #    2.846 CPUs utilized            ( +-  0.38% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              5122      page-faults:u             #    0.155 K/sec                    ( +-  1.23% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      11.646199998 seconds time elapsed                                          ( +-  0.48% )
```

Master, 10 field assignments:

```
 Performance counter stats for 'bash -c yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ./target/release/vector-master-0240ae601 --config ~/tmp/remap.toml > /dev/null' (10 runs):

      36705.275313      task-clock:u (msec)       #    2.968 CPUs utilized            ( +-  0.37% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              5119      page-faults:u             #    0.139 K/sec                    ( +-  0.77% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      12.365062207 seconds time elapsed                                          ( +-  0.39% )
```

This branch, 10 field assignments:

```
 Performance counter stats for 'bash -c yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ./target/release/vector-emit-multiple-3ca72b111 --config ~/tmp/remap.toml > /dev/null' (10 runs):

      49701.886379      task-clock:u (msec)       #    2.876 CPUs utilized            ( +-  0.21% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              4174      page-faults:u             #    0.084 K/sec                    ( +-  0.16% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      17.278687059 seconds time elapsed                                          ( +-  0.33% )
```

One thought I had for not paying that penalty for the empty transform is to delay converting `LogEvent` into a `vrl::Value` until the user actually assigns `.` when the conversion between `vrl::Value` and `event::Value`  would happen anyway.

StephenWakely commented:
> I ran a few benchmarks and found some interesting results:

That is quite the opposite to what I was expecting!

The Vrl Value insert and the Vector Value insert are doing things differently. My first suspicion is that Vector is doing something much more efficient. It will be interesting to compare the two to see what the differences are.

It is a shame that we have two different implementations of exactly the same thing. At some point it would be nice to reconcile the two..

jszwedko commented:
I think the crux of it is converting the `LogEvent` to/from the `vrl::Value` which involves a `BTreeMap` copy in both directions. I might need to rethink this a bit. Maybe I could preserve using `LogEvent` as the target, but, when an array is assigned to `.`, handle that by creating `Vec` of `LogEvent` as the `Target` and handle any array indexing into that struct.

jszwedko commented:
I think this is ready for review now. I updated the implementation and it seems to suffer less overhead (on the order of ~5%; independent of the number fields inserted).

Example:

```toml
[sources.in]
type = "stdin"

[transforms.remap]
type = "remap"
inputs = ["in"]
source = """
[sources.in]
type = "stdin"

[transforms.remap]
type = "remap"
inputs = ["in"]
source = """
. |= object!(parse_json!(.message))
. = .events
"""

[sinks.out]
type = "console"
encoding.codec = "json"
inputs = ["remap"]"""

[sinks.out]
type = "console"
encoding.codec = "json"
inputs = ["remap"]
```

```
➜  vector git:(emit-multiple) ✗ echo '{ "events": [1, "2", true, null, { "message": "hello world"} ]}' | ./target/debug/vector --config /tmp/remap.toml
May 03 20:53:33.057  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info"
May 03 20:53:33.057  INFO vector::app: Loading configs. path=[("/tmp/remap.toml", None)]
May 03 20:53:33.088  INFO vector::sources::stdin: Capturing STDIN.
May 03 20:53:33.088  INFO vector::topology: Running healthchecks.
May 03 20:53:33.089  INFO vector::topology::builder: Healthcheck: Passed.
May 03 20:53:33.089  INFO vector::topology: Starting source. name="in"
May 03 20:53:33.089  INFO vector::topology: Starting transform. name="remap"
May 03 20:53:33.089  INFO vector::topology: Starting sink. name="out"
May 03 20:53:33.090  INFO vector: Vector has started. version="0.14.0" git_version="v0.12.0-289-g90ad54a" released="Mon, 03 May 2021 19:23:17 +0000" arch="x86_64"
May 03 20:53:33.091  INFO vector::shutdown: All sources have finished.
May 03 20:53:33.091  INFO source{component_kind="source" component_name=in component_type=stdin}: vector::sources::stdin: Finished sending.
May 03 20:53:33.091  INFO vector: Vector has stopped.
May 03 20:53:33.093  INFO vector::topology: Shutting down... Waiting on running components. remaining_components="out, remap" time_remaining="59 seconds left"
{"message":1}
{"message":"2"}
{"message":true}
{"message":null}
{"message":"hello world"}
```

Note here I took the suggestion made in #7038 to just use the value as the `message` key for non-object types.

Benchmarks (see above for config):

```
Master empty remap transform:

```
 Performance counter stats for 'bash -c yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ~/bin/vector-master-0240ae601 --config ~/tmp/remap.toml > /dev/null' (10 runs):

      26262.471163      task-clock:u (msec)       #    2.530 CPUs utilized            ( +-  0.59% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              5474      page-faults:u             #    0.208 K/sec                    ( +-  2.26% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      10.381405504 seconds time elapsed                                          ( +-  0.58% )
```

This branch:

```
 Performance counter stats for 'bash -c yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ./target/release/vector-emit-multiple-effc6b3b2 --config ~/tmp/remap.toml > /dev/null' (10 runs):

      29890.939533      task-clock:u (msec)       #    2.723 CPUs utilized            ( +-  0.27% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              5364      page-faults:u             #    0.179 K/sec                    ( +-  2.20% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      10.975345426 seconds time elapsed                                          ( +-  0.27% )
```

Master 10 field inserts:

```
 Performance counter stats for 'bash -c yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ~/bin/vector-master-0240ae601 --config ~/tmp/remap.toml > /dev/null' (10 runs):

      37098.020793      task-clock:u (msec)       #    2.940 CPUs utilized            ( +-  0.33% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              5119      page-faults:u             #    0.138 K/sec                    ( +-  1.25% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      12.618747150 seconds time elapsed                                          ( +-  0.36% )
```

This branch:

```
 Performance counter stats for 'bash -c yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ./target/release/vector-emit-multiple-effc6b3b2 --config ~/tmp/remap.toml > /dev/null' (10 runs):

      38343.198227      task-clock:u (msec)       #    2.920 CPUs utilized            ( +-  2.58% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              5210      page-faults:u             #    0.136 K/sec                    ( +-  1.44% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      13.129863929 seconds time elapsed                                          ( +-  2.59% )
```

I expected a little bit of overhead just creating additional data structures like the iterators, but 5% is more than I expected.

I ran `perf diff` after running `sudo perf record -F 999 -a -g --  ./target/release/vector--config ~/tmp/remap.toml < /tmp/input.txt` on master vs this branch and saw:

```
               +2.53%  vector-emit-multiple      [.] <twox_hash::sixty_four::XxHash64 as core::hash::Hasher>::write
    68.89%     -1.68%  [kernel.kallsyms]         [k] native_safe_halt
               +1.30%  vector-emit-multiple      [.] tracing_subscriber::registry::sharded::Registry::get
               +0.77%  vector-emit-multiple      [.] metrics_tracing_context::TracingContext<R,F>::enhance_key
               +0.71%  vector-emit-multiple      [.] metrics_tracing_context::tracing_integration::MetricsLayer<S>::with_labels
               +0.70%  vector-emit-multiple      [.] dashmap::DashMap<K,V,S>::entry
               +0.58%  vector-emit-multiple      [.] serde_json::ser::format_escaped_str_contents
               +0.53%  vector-emit-multiple      [.] <tracing_subscriber::layer::Layered<L,S> as tracing_core::subscriber::Subscriber>::try_close
               +0.52%  vector-emit-multiple      [.] <tracing_subscriber::registry::sharded::Registry as tracing_core::subscriber::Subscriber>::current_span
               +0.49%  vector-emit-multiple      [.] metrics_tracing_context::TracingContext$LT$R$C$F$GT$::enhance_labels::_$u7b$$u7b$closure$u7d$$u7d$::hf1170eadf40f812f
               +0.49%  vector-emit-multiple      [.] <twox_hash::sixty_four::XxHash64 as core::hash::Hasher>::finish
               +0.48%  vector-emit-multiple      [.] <tracing_subscriber::registry::sharded::Registry as tracing_core::subscriber::Subscriber>::clone_span
               +0.41%  vector-emit-multiple      [.] <metrics::key::KeyData as core::hash::Hash>::hash
               +0.38%  vector-emit-multiple      [.] std::sys_common::backtrace::__rust_begin_short_backtrace
               +0.36%  vector-emit-multiple      [.] futures_channel::mpsc::Receiver<T>::next_message
               +0.35%  vector-emit-multiple      [.] <vector::metrics::recorder::VectorRecorder as metrics::recorder::Recorder>::increment_counter
               +0.23%  vector-emit-multiple      [.] <core::str::lossy::Utf8LossyChunksIter as core::iter::traits::iterator::Iterator>::next
               +0.22%  vector-emit-multiple      [.] vector::event::util::log::insert::map_insert
               +0.22%  vector-emit-multiple      [.] futures_channel::mpsc::Sender<T>::start_send
               +0.21%  vector-emit-multiple      [.] <metrics_tracing_context::TracingContext<R,F> as metrics::recorder::Recorder>::increment_counter
               +0.21%  vector-emit-multiple      [.] <alloc::collections::btree::map::BTreeMap<K,V> as core::clone::Clone>::clone::clone_subtree
               +0.21%  vector-emit-multiple      [.] <vector::event::util::log::path_iter::PathIter as core::iter::traits::iterator::Iterator>::next
     2.57%     +0.20%  [kernel.kallsyms]         [k] __lock_text_start
     1.02%     +0.19%  libc-2.26.so              [.] _int_malloc
     0.61%     +0.19%  libc-2.26.so              [.] cfree@GLIBC_2.2.5
               +0.18%  vector-emit-multiple      [.] regex::re_unicode::Regex::is_match
               +0.16%  vector-emit-multiple      [.] <tracing_subscriber::registry::sharded::Data as tracing_subscriber::registry::SpanData>::extensions
               +0.16%  vector-emit-multiple      [.] chrono::format::format_inner
     1.12%     +0.15%  libc-2.26.so              [.] _int_free
               +0.15%  vector-emit-multiple      [.] <metrics::key::Key as core::cmp::PartialEq>::eq
               +0.13%  vector-emit-multiple      [.] vector::event::value::<impl core::convert::From<vector::event::value::Value> for vrl_compiler::value::Value>::from
               +0.12%  vector-emit-multiple      [.] <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
               +0.12%  vector-emit-multiple      [.] <vector::event::log_event::LogEvent as serde::ser::Serialize>::serialize
               +0.11%  vector-emit-multiple      [.] core::fmt::write
               +0.11%  vector-emit-multiple      [.] <metrics::key::NameParts as core::clone::Clone>::clone
               +0.10%  vector-emit-multiple      [.] core::ptr::drop_in_place<metrics::key::Key>
               +0.09%  vector-emit-multiple      [.] <tracing_subscriber::registry::sharded::CloseGuard as core::ops::drop::Drop>::drop
               +0.09%  vector-emit-multiple      [.] core::ptr::drop_in_place<vrl_compiler::value::Value>
               +0.09%  vector-emit-multiple      [.] alloc::collections::btree::map::BTreeMap<K,V>::insert
               +0.08%  vector-emit-multiple      [.] bytes::bytes::promotable_even_drop
               +0.08%  vector-emit-multiple      [.] alloc::raw_vec::finish_grow
```

`input.txt` was generated via `yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 > /tmp/input.txt`.

The biggest percent differences actually seemed to be at the metric layer (`twox_hash` is used at that layer) which surprised me. I'm open to ideas or thoughts here, but it seems reasonable to me to proceed.

jszwedko commented:
@pablosichert @JeanMertz @StephenWakely do you mind giving this a look?

jszwedko commented on docs/reference/remap.cue:
```diff
@@ -176,6 +176,20 @@ remap: #Remap & {
 			}
 		},
+		{
+			title: "Emitting multiple logs from JSON"
```
I plan to add more interesting examples once the `unnest` function part of #7038 is implemented. I might write up a quick guide too, at that point.

jszwedko replied:
I attempted to add this example, but it didn't quite work out as:

* The VRL tests in `lib/vrl/tests` that test these examples, when testing output, simply add `; .` to the end of the VRL programs and evaluate the return value of executing the program
* The conversion of non-object values to `{ "message": <value> }` actually happens _after_ the VRL program runs, in the `remap` transform

So the example would need to return the primitives, rather than the structured log events, which would be confusing since they are converted to log events.

I think, ideally, the examples in `docs/reference/remap.cue` would be tested at a different level than `lib/vrl/tests` where it would actually run the `remap` transform; including its final conversion. That's a bit more than I'd like to bite off for this PR though so I might defer for now if that's ok. I'll add a test for this to `tests/behavior/transforms/remap.toml` though.

JeanMertz replied:
Yeah, I agree that might be the best solution in the end. However, I think we can still add a test here for documentation purposes, but ignore it in the actual test run itself, at least for now.

jszwedko replied:
👍 I added in 563a627ba

jszwedko commented on lib/vrl/tests/src/docs.rs:
```diff
@@ -37,10 +37,17 @@ pub struct Example {
     #[serde(rename = "return")]
     returns: Option<Value>,
-    output: Option<Event>,
+    output: Option<ExampleOutput>,
```
The test framework needed to be updated to account for the fact that `remap` transforms can emit multiple events now.

jszwedko replied:
Since this is used for deserializing the JSON, I think the enum is needed here? In the `cue export` the `output` field can either be an array or a single element.

jszwedko commented on src/event/log_event.rs:
```diff
@@ -337,50 +341,6 @@ impl Serialize for LogEvent {
     }
 }
 
-impl vrl::Target for LogEvent {
```
Dropped this in lieu of the implementation on `VrlTarget` adapter in `vrl_target.rs`.

jszwedko commented on src/event/log_event.rs:
```diff
@@ -486,256 +446,6 @@ mod test {
         assert_eq!(json.pointer("/map/map/non-existing"), Some(&fallback));
     }
 
-    #[test]
```
Tests were moved to `vrl_target` module.

jszwedko commented on src/event/metric.rs:
```diff
@@ -737,183 +733,6 @@ impl Display for Metric {
-/// fields such as `.tags.host.thing`.
-const MAX_METRIC_PATH_DEPTH: usize = 3;
-
-impl Target for Metric {
```
Dropped in-lieu of implementation on `VrlTarget` adapter in `vrl_target.rs`.

jszwedko commented on src/event/metric.rs:
```diff
@@ -1196,137 +1013,4 @@ mod test {
         );
     }
-
-    #[test]
```
Moved to `vrl_target`.

jszwedko commented on src/event/vrl_target.rs:
```diff
@@ -0,0 +1,740 @@
+
+/// An adapter to turn `Event`s into `vrl::Target`s.
+#[derive(Debug, Clone)]
+pub enum VrlTarget {
```
This is the bulk of this PR. It introduces an adapter for `Event` to satisfy `vrl::Target` to allow, for log events, `.` to be set to something other than a map (for example, here we want to allow it to be set to an array).

JeanMertz commented on src/transforms/remap.rs:
```diff
@@ -87,22 +87,27 @@ impl FunctionTransform for Remap {
                 emit!(RemapMappingAbort {
                     event_dropped: self.drop_on_abort,
                 });
 
                 if !self.drop_on_abort {
-                    output.push(original_event.unwrap_or(event))
+                    if let Some(event) = original_event {
+                        output.push(event)
+                    }
```
Hmm, doesn't this change the original behavior, given that we no longer default to using `event`?

jszwedko replied:
Apologies, I meant to call this change out as it is subtle. The original behavior using `unwrap_or` would never actually call the `unwrap_or` if it got this far as if `drop_on_abort` is `false`, and the program _can_ abort, then `original_event` is always `Some`. If `drop_on_abort` is `false` but the program _cannot_ abort, then this `Err(Terminate::Abort)` match arm will never be hit.

It would be nice to model this a bit more clearly. I'll see if I can think of something.

jszwedko replied:
I couldn't come up with anything better at the moment so I just swapped the `if let Some`s for `expect` to at least make the invariant more clear.

JeanMertz replied:
Ah yep, good catch. The `expect` solution seems fine to me.

JeanMertz reviewed (changes requested):
Nice! I really dig the cleanness of this solution. I left a few comments and one blocking question, but other than that this looks good to me 👍

jszwedko commented on src/event/vrl_target.rs:
```diff
@@ -0,0 +1,740 @@
+// * If `.` is an array, map over all of the values to create log events:
+//   * If an element is an object, create an event using that as fields
+//   * If an element is anything else, assign to the `message` key
+// * If `.` is anything else, assign to the `message` key
```
👍 this is a change in behavior. The current behavior is actually just to ignore the assignment. Good call on documentation, though I'll add.

jszwedko commented:
@JeanMertz thanks for the feedback! I think I resolved everything.

StephenWakely commented on lib/vector-core/src/event/vrl_target.rs:
```diff
@@ -0,0 +1,750 @@
+}
+
+impl vrl_core::Target for VrlTarget {
+    fn insert(&mut self, path: &LookupBuf, value: vrl_core::Value) -> Result<(), String> {
```
Perhaps it's out of scope for this, but I wonder if instead of cloning the event at the start of the remap transform we could move the clone to here.

`VrlTarget` could be something like:

```rust
pub struct VrlTarget {
    retain_original: boolean,
    originalEvent: Option<BTreeMap<String, Value>>,
    event: enum {
        LogEvent(Value, EventMetadata),
        Metric(Metric),
    } 
}
```

Then here we could:

```rust
  if self.retain_original && self.original.is_none() {
    self.original = self.value.as_map().clone();
  }
```

This could have a significant boost for scripts that abort the transaction prior to modifying any fields..

jszwedko replied:
🤔 that's true. I feel like I'd prefer to punt on that for now since this maintains the current status quo. I think, ideally, the whole event would be "copy-on-write" so we wouldn't need to worry about the clone cost here.

StephenWakely reviewed (approved):
Looks good to me!

Is there a reason why we aren't allowing it for Metrics beyond it would be more complicated to implement and would likely never be used?

jszwedko commented:
> Is there a reason why we aren't allowing it for Metrics beyond it would be more complicated to implement and would likely never be used?

I feel like metric support with remap is still a little rudimentary. It's easy to do things like modify the tags on a metric, but creating a metric via assigning `.` would be tricky. I think they'd need to set it to a `vrl::Value::Map` with the known keys (`name`, `tags`, etc.) and then we'd need to validate when marshaling out that it matches the "schema" and create a `event::Metric` event from it. Or maybe use a special `create_metric` function that can have required / optional arguments. 

I think the log event case is more common though so I figured I'd start with that. We can see if people ask for support for multiple metrics too.

pablosichert commented on lib/vector-core/src/event/vrl_target.rs:
```diff
@@ -0,0 +1,751 @@
+/// We can get the `type` of the metric in Remap, but can't set it.
+const VALID_METRIC_PATHS_GET: &str = ".name, .namespace, .timestamp, .kind, .tags, .type";
+
+/// Metrics aren't interested in paths that have a length longer than 3
```
```suggestion
/// Metrics aren't interested in paths that have a length longer than 3.
```

pablosichert commented on lib/vector-core/src/event/vrl_target.rs:
```diff
@@ -0,0 +1,751 @@
+/// An adapter to turn `Event`s into `vrl_core::Target`s.
+#[derive(Debug, Clone)]
+pub enum VrlTarget {
+    // LogEvent is essentially just a destructured `event::LogEvent`, but without the semantics
+    // that `fields` must always be a Map variant
```
```suggestion
    // LogEvent is essentially just a destructured `event::LogEvent`, but without the semantics
    // that `fields` must always be a `Map` variant.
```

pablosichert commented on lib/vector-core/src/event/vrl_target.rs:
```diff
@@ -0,0 +1,751 @@
+}
+
+// Turn a Value back into LogEvents:
+// * In the common case, where `.` is an map, just create an event using it as fields
```
```suggestion
// * In the common case, where `.` is a map, just create an event using its fields
```

pablosichert commented on lib/vector-core/src/event/vrl_target.rs:
```diff
@@ -0,0 +1,751 @@
+    }
+}
+
+// Turn a Value back into LogEvents:
```
```suggestion
// Turn a `Value` back into `LogEvents`:
```

pablosichert reviewed (approved):
Nice work, @jszwedko!

There still seems to be a small clippy error after merging with master.

I also left some pedantic wording/punctuation suggestions.

jszwedko commented on docs/reference/components/transforms/remap.cue:
```diff
@@ -138,6 +138,17 @@ components: transforms: "remap": {
+				to the root path `.`. One log event will be emitted for each input
+				element of the array.
+
+				If any of the array elements is not an object, they will be set as
+				the `message` key for the corresponding emitted log event.
```
👍 I ended up rewriting it as:

```
                If any of the array elements is not an object, a log event will
                be created that uses the element value as the `message` key. For
                example, `123` will be emitted as `{ "message": 123 }`
```
                
I agree that the original was confusing.

jszwedko commented:
@JeanMertz I'll merge this given the two approvals and that I think I addressed all of your comments, but let me know if you see anything else and I can address in a follow-up.
